File: rpi/mqtt_client.py
import paho.mqtt.client as mqtt
import logging

def on_connect(client, userdata, flags, rc):
    logging.info("MQTT - connected to broker, code " + str(rc))


def on_message(client, userdata, msg):
    global message
    message = msg
    topic = msg.topic
    payload = str(msg.payload.decode("utf-8"))
    logging.debug("MQTT - msg received [" + topic + "]: " + payload)


def on_subscribe(client, userdata, mid, granted_qos):
    logging.info("MQTT - subscribed: " + str(mid) + " " + str(granted_qos))


def on_publish(client, userdata, mid):
    logging.debug("MQTT - published:  " + str(mid))


def on_log(client, userdata, level, string):
    logging.debug(string)


def connect(ip, port=1883):
    mqttc = mqtt.Client()
    mqttc.on_connect = on_connect
    mqttc.on_message = on_message
    mqttc.on_publish = on_publish
    mqttc.on_subscribe = on_subscribe
    # mqttc.on_log = on_log
    try:
        mqttc.connect(ip, port, 60) 
    except:
        logging.error("MQTT - [" + ip + ":" + str(port) + "] connection failed!, exiting...")
        exit(1)
    return mqttc


if __name__ == "__main__":
    mqtt_broker_ip = "192.168.1.3"
    mqtt_broker_port = 1883
    mqtt_topic = "testMQTT"
    mqttc = connect(mqtt_broker_ip, mqtt_broker_port)
    mqttc.subscribe(mqtt_topic)
    mqttc.loop_forever()

File: microcontroller/MqttClient/MqttClient.cpp
#include "MqttClient.h"
#include <MQTT.h>
#include <ESP8266WiFi.h>

MqttClient::MqttClient(WiFiClient wifi_client, const char *mqtt_broker_ip, const int mqtt_broker_port)
{
  _mqtt_client.begin(mqtt_broker_ip, wifi_client);
}

void callback(String &topic, String &payload)
{
  Serial.println("Message arrived [" + topic + "]: " + payload);
  _msg = payload;
  _topic = topic;
}

void MqttClient::connect(const char *client_id, String *topics, int num_topics)
{
  Serial.print("Connecting to MQTT broker... ");
  while (!_mqtt_client.connect(client_id, "public", "public"))
  {
    Serial.print(".");
    delay(1000);
  }
  Serial.println("done!");
  for (int i = 0; i < num_topics; i++)
  {
    _mqtt_client.onMessage(callback);
    _mqtt_client.subscribe(topics[i]);
    Serial.println("Subscribed to topic: " + topics[i]);
  }
}

void MqttClient::publish_message(char *topic, char *msg)
{
  _mqtt_client.publish(topic, msg);
  Serial.println("Message published [ " + String(topic) + " ]: " + String(msg));
}

void MqttClient::check_connection(const char *client_id, String *topics, int num_topics)
{
  _mqtt_client.loop();
  delay(10);
  if (!_mqtt_client.connected())
    connect(client_id, topics, num_topics);
}

String MqttClient::get_msg() { return _msg; }

String MqttClient::get_topic() { return _topic; }

void MqttClient::reset_msg()
{
  _msg = "";
  _topic = "";
}
